### DISCUSSION TITLE
Lost sequence numbers of idle frames in VcTmPacketHandler class

### DISCUSSION THREAD

**Discussion by fgretief** [yamcs/yamcs#598] [General]
Hi,

I came across an issue in the [`VcTmPacketHandler`](https://github.com/yamcs/yamcs/blob/0fadce4070fa696b08441789ee32a0f58e4ef729/yamcs-core/src/main/java/org/yamcs/tctm/ccsds/VcTmPacketHandler.java#L79) class while trying to debug missing TmTransferFrames on a lossless interface.

In the `handle()` method, there is a frameloss calculation that depends on the `lastFrameSeq` state variable which is updated everytime a packet is being processed. Higher up in that method, there is an if statement that drops idle frames.  Due to the dropping, the frameloss calculation continuesly indicates that there were lost frame when in fact there were none. The idle frames contain the correct sequence numbers. 

To fix the issue, I changed the code to update the `lastFrameSeq` counter with the idle frame's sequence counter when it is dropped.  

Is this the best way to do it or am I missing something else?

Commit here: https://github.com/fgretief/yamcs/commit/adcf741e0908d4c19ba361c8dbbc39e9016e2cd0

**Comment by fgretief** [comment]
On a related note: Why is the value that the `getSeqCountWrapArround()` method in `TmTransferFrame` class equal to 255? Should it not be 256?

In the case of TmTransferFrames, the sequence counter is 8-bits wide. Thus 0 to 255. The sequence counter increments sequentially, so roll-over happens from 255 to 0.  In the `lostFramesCount()` method, the value of `delta` during a roll-over is calculated as `vcFrameSeq + getSeqCountWrapAround() - prevFrameSeq`. [code](https://github.com/yamcs/yamcs/blob/0fadce4070fa696b08441789ee32a0f58e4ef729/yamcs-core/src/main/java/org/yamcs/tctm/ccsds/DownlinkTransferFrame.java#L93)

For roll-over the `vcFrameSeq` is 0, `prevFrameSeq` is 255 and `getSeqCountWrapAround()` returning 255 results in `delta` being 0. Then a few lines lower, `delta` is decreased to -1, which is the value being returned by `lostFrameCount()`.

But this is wrong, the value should be 0 for a roll-over because no packets have been lost. Also, in the `handler()` method that uses this value, the `frameloss != 0` check causes the `"Incomplete packet dropped because of frame loss"` message to be printed just before the decoder is reset, causing a perfectly good packet to be ignored.

So, to my understaning of the situation, if the `getSeqCountWrapAround()` would return 256 instead of 255, the result of the decremented `delta` would be 0 (as expected) which causes the rest of the code to work correctly. 

I have only checked/tested with TmFrames, but AosTransferFrame looks like it might have the same issue. 

Alternatively, there could be a +1 added to the `delta` calculation for the roll-over case, in stead of changing the value of `getSeqCountWrapAround()` methods.

**Comment by xpromache** [comment]
Thanks for your messages, you seem to be right on both issues, will you
make a pull request?

1.  update the lastFrameSeq counter with the idle frame's sequence counter
when it is dropped.
2. change the return of getSeqCountWrapAround() (and also for AOS and USLP
please).

thx again!

**Comment by xpromache** [comment]
On a second thought maybe is better just to add  a +1 on that line:
``` 
long delta = prevFrameSeq < vcFrameSeq ? 
             vcFrameSeq - prevFrameSeq 
           :   vcFrameSeq + getSeqCountWrapArround() - prevFrameSeq + 1;

```

what do you think?

**Reply by fgretief** [reply]
Yes, I think this is the more generic solution.

**Comment by fgretief** [comment]
Pull request #599 submitted.
